Code/GraphMol/Descriptors/WHIM.cpp

```diff
@@ -0,0 +1,625 @@
+
+    MatrixXd WeigthMat = Weigth.asDiagonal();
+
+
+    double weigth=WeigthMat.diagonal().sum();
+
+
+    MatrixXd Xmean = GetCenterMatrix(MatOrigin);
+
+    MatrixXd covmat = GetCovMatrix(Xmean,WeigthMat,weigth);
+
+    JacobiSVD<MatrixXd> svd = getSVD(covmat);
+
+    w= getWhimDesc(svd, Xmean, numAtoms, th, false);
+
+
+
+    return w;
+  }
+
+
+
+double*  GetWHIMIonPol(const Conformer &conf, double Vpoints[], double th){
+      double *w = new double[18];
+
+    int numAtoms = conf.getNumAtoms();
+
+    Map<MatrixXd> matorigin(Vpoints, 3,numAtoms);
+
+    MatrixXd MatOrigin=matorigin.transpose();
+
+    std::vector<double> weigthvector = moldata3D.GetRelativeIonPol(conf.getOwningMol());
+
+    double* weigtharray = &weigthvector[0];
+
+    Map<VectorXd> Weigth(weigtharray,numAtoms);
+
+    MatrixXd WeigthMat = Weigth.asDiagonal();
+
+
+    double weigth=WeigthMat.diagonal().sum();
+
+
+    MatrixXd Xmean = GetCenterMatrix(MatOrigin);
+
+    MatrixXd covmat = GetCovMatrix(Xmean,WeigthMat,weigth);
+
+    JacobiSVD<MatrixXd> svd = getSVD(covmat);
+
+    w= getWhimDesc(svd, Xmean, numAtoms, th,false);
+
+
+
+    return w;
+  }
+
+
+double*  GetWHIMIState(const Conformer &conf, double Vpoints[], double th){
+    double *w = new double[18];
+
+    ROMol& mol = conf.getOwningMol();
+
+    MolOps::removeHs(mol, false, false);
+
+    int numAtoms = mol.getNumAtoms();
+
+    Map<MatrixXd> matorigin(Vpoints, 3,numAtoms);
+
+    MatrixXd MatOrigin=matorigin.transpose();
+
+    std::vector<double> weigthvector = moldata3D.GetIState(mol);
+
+    double* weigtharray = &weigthvector[0];
+
+    Map<VectorXd> Weigth(weigtharray,numAtoms);
+
+    MatrixXd WeigthMat = Weigth.asDiagonal();
+
+
+    double weigth=WeigthMat.diagonal().sum();
+
+
+    MatrixXd Xmean = GetCenterMatrix(MatOrigin);
+
+    MatrixXd covmat = GetCovMatrix(Xmean,WeigthMat,weigth);
+
+    JacobiSVD<MatrixXd> svd = getSVD(covmat);
+
+    w= getWhimDesc(svd, Xmean, numAtoms, th,false);
+
+    return w;
+  }
+
+
+
+} //end of anonymous namespace
+
+
+
+
+std::vector<double> WHIM(const ROMol& mol,int confId, double th){
+  PRECONDITION(mol.getNumConformers()>=1,"molecule has no conformers")
+  int numAtoms = mol.getNumAtoms();
+
+  const Conformer &conf = mol.getConformer(confId);
+
+  double Vpoints[3*numAtoms];
+
+  for(int i=0; i<numAtoms; ++i){
+     Vpoints[3*i]   =conf.getAtomPos(i).x;
+     Vpoints[3*i+1] =conf.getAtomPos(i).y;
+     Vpoints[3*i+2] =conf.getAtomPos(i).z;
+  }
+
+  std::vector<double> res;
+
+  double* wu= GetWHIMU(conf, Vpoints, th);
+  double* wm= GetWHIMMass(conf, Vpoints, th);
+  double* wv= GetWHIMvdw(conf, Vpoints, th);
+  double* we= GetWHIMneg(conf, Vpoints, th);
+  double* wp= GetWHIMpol(conf, Vpoints, th);
+  double* wi= GetWHIMIonPol(conf, Vpoints, th);
+  double* ws= GetWHIMIState(conf, Vpoints, th);
+
+
+  // takes only L1u L2u L3u P1u P2u G1u G2u G3u E1u E2u E3u
+  int map1[11] = {0,1,2,6,7,14,15,16,10,11,12};
```

Comment:
it's the definition order in Dragon, again for comparison, it's better to have the same order but not necessary if we can look up using descriptor names. WHIM is not the easiest descriptor by the way. This is where we have the more divergence between both codes (this implementation vs Dragon) related to G values that used symmetry of the atom position... but this is not trivial to acheive.